vhdl_lang/src/analysis/declarative.rs

```diff
@@ -6,21 +6,94 @@
 
 use super::names::*;
 use super::*;
-use crate::ast;
 use crate::ast::*;
 use crate::data::*;
 use crate::named_entity::{Signature, *};
+use crate::{ast, HasTokenSpan};
 use analyze::*;
 use fnv::FnvHashMap;
 use std::collections::hash_map::Entry;
 
+/// Describes where a declaration was made
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum DeclarativeContext {
+    Block,
+    Configuration,
+    Entity,
+    PackageBody,
+    Package,
+    Process,
+    ProtectedTypeBody,
+    Subprogram,
+}
+
+impl Declaration {
+    pub fn is_allowed_in_context(&self, ctx: DeclarativeContext) -> bool {
+        use Declaration::*;
+        use ObjectClass::*;
+        match ctx {
+            DeclarativeContext::Block => true,
+            DeclarativeContext::Configuration => {
+                matches!(self, Use(_) | Attribute(ast::Attribute::Specification(_)))
+            }
+            DeclarativeContext::Entity => matches!(
+                self,
+                Object(_)
+                    | File(_)
+                    | Type(_)
+                    | Attribute(_)
+                    | Alias(_)
+                    | SubprogramDeclaration(_)
+                    | SubprogramInstantiation(_)
+                    | SubprogramBody(_)
+                    | Use(_)
+                    | Package(_)
+            ),
+            DeclarativeContext::PackageBody
+            | DeclarativeContext::Process
+            | DeclarativeContext::ProtectedTypeBody
+            | DeclarativeContext::Subprogram => matches!(
+                self,
+                Object(ObjectDeclaration {
+                    class: Constant | Variable,
+                    ..
+                }) | File(_)
+                    | Type(_)
+                    | Attribute(_)
+                    | Alias(_)
+                    | SubprogramDeclaration(_)
+                    | SubprogramInstantiation(_)
+                    | SubprogramBody(_)
+                    | Use(_)
+                    | Package(_)
+            ),
+            DeclarativeContext::Package => matches!(
+                self,
+                Object(ObjectDeclaration {
+                    class: Constant | Variable,
+                    ..
+                }) | File(_)
+                    | Type(_)
+                    | Component(_)
+                    | Attribute(_)
+                    | Alias(_)
+                    | SubprogramDeclaration(_)
+                    | SubprogramInstantiation(_)
+                    | Use(_)
+                    | Package(_)
+            ),
+        }
+    }
+}
+
 impl<'a> AnalyzeContext<'a> {
     pub fn analyze_declarative_part(
         &self,
         scope: &Scope<'a>,
         parent: EntRef<'a>,
         declarations: &mut [Declaration],
         diagnostics: &mut dyn DiagnosticHandler,
+        context: DeclarativeContext,
```

Comment:
Suggestion: The `parent` argument to the function is kidn of redundant with the new DeclarativeContext you added. The parent will be the package, block or whatever is the owner of the declarative part. So this code could potentially be simplified.

Reply:
This is not always true though. For example, a `Block` or a `Process` simply passes the `parent` to the declarative part https://github.com/VHDL-LS/rust_hdl/blob/master/vhdl_lang/src/analysis/concurrent.rs#L99

Reply:
A block or process do not always define a label. Should they still be named entities?

Reply:
The code is a bit complicated for concurrent and sequential statements since labels need to be visible before they are defined. Thus there is a bit of separation between the code which creates the AnyEnt for the block and the code which analyses the declarative part of the block. But I still think the parent is correct in this scenario. Otherwise the document symbols (outline in vscode) would be wrong.

Reply:
Can you point out to me where the `AnyEnt` for the block is created? I only see the parent... (I pasted the wring line; it should be 103, sorry)

Reply:
I think they become named entities with Designator::Anonymous which I added just for DocumentSymbols. I think the anonymous labels even comes from the LRM but I am not sure and cannot check it now.

Reply:
The AnyEnt for the block is created in the self.define_labels_for_concurrent_part for the surrounding block.

Reply:
You can view this as an optional improvement suggestion from me that you should look into a bit. If it turns out that it is not as straightforward as I imagined we can keep your current solution with the extra argument.

As long as the example project works it is ok to merge.

Reply:
I'll do it. I just didn't see that block was also always the correct `EnyEnt`. Example project now parses fine